### ISSUE TITLE
Improve handling of the named pipe used by `netdatacli`.

### ISSUE THREAD

**Issue by Ferroin** [netdata/netdata#7587]
Currently, the named pipe that is used for sending commands to the agent process from the `netdatacli` command is hardcoded to be located at `/tmp/netdata-ipc` on UNIX systems. This is problematic for a few reasons:

* `/tmp` is world writable. This means that a user can theoretically cause issues by creating a file (or symlink) with the same name while Netdata is not running.
* `/tmp` is not consistently 'unified'. In some setups, each user has their own view of this directory, which 
* Per a strict inteprretation of the filesystem hierarchy standards, `/tmp` is not the correct place for the named pipe anyway, since it's a runtime file and not a temporary file.

To resolve these issues, I would propose the following two changes:

* The default location for this named pipe on UNIX systems should be changed to either `/run` (on Linux and other systems which provide `/run`) or `/var/run` (on systems which do not provide `/run`). This will solve the first and third issues above, but not the second one (which is also true for `/run`)
* The exact location should be made user configurable with an option in the config file. This will solve the second issue above.

**Comment by mfundul** [comment]
I picked `/tmp/` since this is the default location for `mysql` and `postgreSQL` domain sockets as well. Moreover, I needed a directory that is guaranteed to not be globally viewable. E.g. if there are many containers running netdata inside, each container should have its own private socket to interface with its own private cli tool, which I hope is guaranteed with the choice of `/tmp/`.

Let's try not to break this assumption.

**Comment by mfundul** [comment]
Then of course there is the question about the official docker installation of netdata. Are we supposed to separately install cli in the host filesystem and the rest in the container?

**Comment by amoss** [comment]
It is true that the socket is in the wrong directory. I think that /run or /var/run matches the semantics better and we can create e.g. /var/run/netdata/ at install-time to  solve problem (1) above.

Our docker image does not bind-mount /run or /var/run so each container would still have its own private socket. I don’t see the problem with multiple containers that @mfundul describes above.

Problem (2) sounds like a benefit - this means that the netdata user would effectively have a private /tmp directory.

Number (3) is correct and it could cause a problem for our district packaging.

**Comment by Ferroin** [comment]
> It is true that the socket is in the wrong directory. I think that /run or /var/run matches the semantics better and we can create e.g. /var/run/netdata/ at install-time to solve problem (1) above.

Creating a directory in `/var/run` for it won't really help unless we do it at runtime, at least not on any modern Linux system (almost all modern Linux systems have `/var/run` as a symlink pointing to `/run`, which means the contents are lost on reboot (because `/run` is pretty much always a tmpfs mount).

> Problem (2) sounds like a benefit - this means that the netdata user would effectively have a private /tmp directory.

It is, but not for this. If the `netdata` user has a private `/tmp`, then nobody can use `netdatacli` unless they're logged in as the `netdata` user, because everyone else will see a different view of `/tmp` (and marking `netdatacli` SUID as the `netdata` user would not solve this).

**Comment by mfundul** [comment]
I think currently only the `netdata` user can actually use `netdatacli`.

**Comment by Ferroin** [comment]
@mfundul It works for both root and whatever user `netdata` is running as. I think in some cases it might also work for people in the `netdata` group, but I'm not certain on that.

**Comment by mfundul** [comment]
> 
> 
> @mfundul It works for both root and whatever user `netdata` is running as. I think in some cases it might also work for people in the `netdata` group, but I'm not certain on that.

Yes, the idea was it should work for the configured group. So, the risk currently is that the root user could lose access if `/tmp/` is somehow only available to `netdata` user?

**Comment by Ferroin** [comment]
> > @mfundul It works for both root and whatever user `netdata` is running as. I think in some cases it might also work for people in the `netdata` group, but I'm not certain on that.
> 
> Yes, the idea was it should work for the configured group. So, the risk currently is that the root user could lose access if `/tmp/` is somehow only available to `netdata` user?

Not just root, but in theory anyone in the group too.

In general, there's three ways that private `/tmp` can be implemented:

* Per-user sub-directories in `/tmp` (usually hidden by prefixing with a `.`) and a custom value for `$TMPDIR` in the environment. In this case SUID won't get you the right `/tmp`, but `sudo` _might_ (depends on what environment variables it pulls in).
* Per-user mount namespaces, with `/tmp` being part of the mount namespace, often stacked on top of the per-user option (the per-user directory gets bind-mounted in the mount namespace). Here, neither SUID nor `sudo` will get you the right `/tmp`, you have to be in the specific login session to get it.
* LSM configuration hiding things not owned by the user in `/tmp` (SELinux can be configured to do this). Here, SUID may or may not work, but `sudo` usually will work.

In contrast, `/run` is almost never namespaced like that (though some LSM's may be configured to hide it from regular users).

**Comment by prologic** [comment]
Can we just make this configurable already? It seems rather limiting to have to hard-code something like this, most software (_MySQL, PostgrSQL included_) I'm fairly sure make this configurable! Making this configurable lets us be a bit more flexible at the packaging/systems level which would help to solve #8898

**Comment by mfundul** [comment]
> 
> 
> Can we just make this configurable already? It seems rather limiting to have to hard-code something like this, most software (_MySQL, PostgrSQL included_) I'm fairly sure make this configurable! Making this configurable lets us be a bit more flexible at the packaging/systems level which would help to solve #8898

We should.

**Comment by amoss** [comment]
> Making this configurable lets us be a bit more flexible at the packaging/systems level which would help to solve #8898

How does making the named pipe location configurable help fix #8898? I thought that required  change to the location of the netdata `.pid` file.

**Comment by prologic** [comment]
> > Making this configurable lets us be a bit more flexible at the packaging/systems level which would help to solve #8898
> 
> How does making the named pipe location configurable help fix #8898? I thought that required change to the location of the netdata `.pid` file.

Its a bit of both I believe. Since we want to use `netdatacli` more to reliably talk to the actual daemon running and tell it to "do stuff" we also need to ensure we can properly talk to it and not assume we know where its UNIX socket is. The pid file has always been configurable, but our shutdown sequence tries to (_since some tiem ago_) use `netdatacli shutdown-agent` preferentially I believe.

**Comment by Ferroin** [comment]
I’m planning to pick this up some time in the near future as it’s blocking a couple of other issues and should (in theory) be an easy fix.

**Comment by OdysLam** [comment]
This is also relevant to #10721, which appears intermitently.

**Comment by wmertens** [comment]
On NixOS, mariadb is configured like so:

https://github.com/NixOS/nixpkgs/blob/3b6c3bee9174dfe56fd0e586449457467abe7116/pkgs/servers/sql/mysql/8.0.x.nix#L46

and mariadb takes care of creating the missing directories.

See also https://github.com/NixOS/nixpkgs/pull/127382

**Comment by ilyam8** [comment]
@stelfrag @Ferroin do we still need this issue? Do someone plan to fix it?

**Comment by Ferroin** [comment]
Yes, we still need this, because nothing has changed and the issues I outlined with the current approach in the original post are _still issues_. I’m not sure if anybody is actually planning to work on it currently (back when i said I would, I as usual got call off to work on some other thing that was ‘top priority’ and never got back to this). Nobody currently planning to work on it does not mean it’s not an issue though, and definitely does not mean it should be closed.

**Comment by ilyam8** [comment]
1. _Outlines_ issues are very very weak and almost non-existent: theoretically..., not unified, doesn't satisfy a strict interpretation of the filesystem hierarchy standards.
2. It's been (sic!!!) 5 years and as you admit yourself - nobody is and will work on it.

For me this 100% falls into the "won't be fixed" category (yes, it won't be fixed). It doesn't matter if that is an issue or not for you, it is effectively "wontfix" from us. And let me repeat - 5 years, no movement.
